File: assets/src/codecon18/xor_cypher.py
import sys

def decode(msg, key):
  return ''.join([
      decode_digit(key, msg[i:i+2])
      for i in range(0, len(msg), 2)
    ]);
   
def decode_digit(key, doubleHexDigit):
  return chr(key ^ int(doubleHexDigit, 16));
  
'''
def my_decode(msg, key): return ''.join([chr(key ^ int(msg[i:i+2], 16)) for i in range(0, len(msg), 2)])
'''  

def evaluate_credibility(str):
  tokens = str.split()
  words = [ word for token in tokens for word in token.split("-") ]
  return sum(score_words(words))
  
def score_words(words):
   for w in words:
      if w.isalnum(): 
        if len(w) > 1 and w[0].isupper() and w[-1].islower():
          yield 25
        else:
          yield 5
          
def solve(input):
  messages = [decode(input, key) for key in range(128)]
  return max(messages, key=evaluate_credibility)
  
data = sys.stdin.read().splitlines()
for line in data:
    print (solve(line))
    
''' 
    print([(str(evaluate_credibility(msg)), msg) for msg in messages])
'''

File: assets/src/codecon18/xor_cypher.hs
import Data.Char (digitToInt, chr, isAlphaNum, isLower, isUpper)
import Data.Bits (xor)
import Data.Text (splitOn, pack, unpack)
import Test.QuickCheck.Text(oneLine)
import Data.Ord (comparing)
import Data.List (maximumBy)

-- For an interactive use: type main, then type the input and end it with Ctrl+Z on Windows / Ctrl+D in a Bash Terminal
main :: IO ()
main = interact $ show . solve . oneLine

decode :: Int -> String -> String
decode key (a:b:xs) = decode_digit key a b : decode key xs
decode _ _ = []

decode_digit :: Int -> Char -> Char -> Char
decode_digit key a b = chr (xor key (16*digitToInt a + digitToInt b))

-- decode key (a:b:xs) = chr (xor key (16*digitToInt a + digitToInt b)) : decode key xs
            
splitWords :: String -> [String]
splitWords msg = concatMap words (map unpack (splitOn (pack "-") (pack msg)))

evalCred  :: String -> Integer
evalCred s = foldl sumScore 0 (splitWords s)
    where sumScore acc word = acc + 
            if all isAlphaNum word 
                then if isUpper (head word) && isLower (last word)
                    then 25
                    else 5
                else 0
                
-- Check it out like this:
-- solve "023f6c386b3f39222820326b3f246b392428206b2a6b392332262e6b3f246b392428206b2a6b392332262e6b3f232a3f6c386b39222c233f6b24256b3f22262e"
solve :: String -> String
solve s = maximumBy (comparing evalCred) (map (\key -> decode key s) [0..127])
